### PR TITLE
teku: fix build

### PR DESCRIPTION
**Pull request by ajsutton**
The Teku build process determines the version using git tags, but the Formula was building from a downloaded source tarball which does not include tag information. As a result, `teku --version` would report `UNKNOWN` as the version rather than the actual version built.

Additionally, the Teku build does not support arbitrary versions of gradle, and should always be built with the `./gradlew` wrapper script to ensure the correct version of gradle is used.

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by SMillerDev** [comment]
> The Teku build process determines the version using git tags, but the Formula was building from a downloaded source tarball which does not include tag information. As a result, teku --version would report UNKNOWN as the version rather than the actual version built.

Is there a variable we can pass to the build to fix this? Because a tarbal download is much quicker than a git checkout.

> Additionally, the Teku build does not support arbitrary versions of gradle, and should always be built with the ./gradlew wrapper script to ensure the correct version of gradle is used.

Is it actually broken in the current form? Because we want to avoid using vendored versions of homebrew formulae (which the gradle wrapper will download without checksum validation)

**Comment by ajsutton** [comment]
> Is there a variable we can pass to the build to fix this? Because a tarbal download is much quicker than a git checkout.

No there isn't.  The build process is only designed to work from a git checkout.

> Is it actually broken in the current form? Because we want to avoid using vendored versions of homebrew formulae (which the gradle wrapper will download without checksum validation)

We are using the current version of Gradle but regularly see issues with Gradle upgrades that then require various plugins to be updated before they work. So once Homebrew updates to the next Gradle release, there's a good chance that we won't be able to update Teku until we've done the work to support that Gradle release. Given that timely updates for Ethereum clients is important both to stay current with network upgrades and respond to potential security threats that worries me a great deal.

One of my concerns here is that currently we have a homebrew tap we provide for Teku which uses official builds that were built and tested in our CI. Putting this into homebrew means we're doing a different build of Teku which isn't undergoing any testing and then likely becomes the default version for all Mac users.  The more we diverge the build process in homebrew from what we do in CI, the more likely we are to introduce problems. Most Teku users will have it controlling over $100,000 USD worth of staked ETH so quality control is particularly paramount here, and downtime for them can be quite expensive.

**Comment by SMillerDev** [comment]
It's open source, so not everyone is going to build it exactly how upstream suggests. https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#official-vendor-taps has some notes on this.

If people want the exact version you build they can use the tap, and if they want a version that's more in line with the homebrew standards they can use homebrew-core. Seems like a good solution to me.

**Comment by ajsutton** [comment]
> It's open source, so not everyone is going to build it exactly how upstream suggests. https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#official-vendor-taps has some notes on this.

I think the concern here isn't that it's not built exactly how upstream suggests, but that it's being built in a way that means homebrew has to commit to delaying updates to Gradle until Teku is able to build with newer versions or becomes unable to build and ship updates to Teku until it works with the newer gradle. Neither seem like a very good outcome.

> If people want the exact version you build they can use the tap, and if they want a version that's more in line with the homebrew standards they can use homebrew-core. Seems like a good solution to me.

Currently Homebrew-core is shipping a version that reports its version as `UNKNOWN` which is a bit of a problem from a support perspective.

So coming back to your original concern:
> Because we want to avoid using vendored versions of homebrew formulae (which the gradle wrapper will download without checksum validation)

I see that it's possible to enable checksum validation in the Gradle wrapper which would be a good improvement for us to add upstream. Would that resolve your concerns? Then the specific version of gradle would be downloaded just like all the other Teku dependencies are as part of the build.

**Comment by SMillerDev** [comment]
> I think the concern here isn't that it's not built exactly how upstream suggests, but that it's being built in a way that means homebrew has to commit to delaying updates to Gradle until Teku is able to build with newer versions or becomes unable to build and ship updates to Teku until it works with the newer gradle. Neither seem like a very good outcome.

For the latter we can ship an older gradle version alongside but in general yes, this is how package management works.

> Currently Homebrew-core is shipping a version that reports its version as UNKNOWN which is a bit of a problem from a support perspective.

I have no objections to that change

> I see that it's possible to enable checksum validation in the Gradle wrapper which would be a good improvement for us to add upstream. Would that resolve your concerns? Then the specific version of gradle would be downloaded just like all the other Teku dependencies are as part of the build.

That would probably be good for upstream, but that would still download a version of software that Homebrew already ships. Unless something is actually broken or causing problems I think we should stay with the Homebrew build Gradle.

**Comment by ajsutton** [comment]
> That would probably be good for upstream, but that would still download a version of software that Homebrew already ships. Unless something is actually broken or causing problems I think we should stay with the Homebrew build Gradle.

Well this goes against Gradle's own recommendations and will almost certainly lead to problems on the next major gradle update, but as long as the homebrew team are happy to take the burden of working that out then so be it.  I've reverted the use of the gradlew wrapper for this PR.

File: Formula/teku.rb [hunk 1]
```diff
@@ -1,9 +1,9 @@
 class Teku < Formula
   desc "Java Implementation of the Ethereum 2.0 Beacon Chain"
   homepage "https://docs.teku.consensys.net/"
-  url "https://github.com/ConsenSys/teku/archive/refs/tags/21.12.2.tar.gz"
-  sha256 "fdd2e23ee8228b0bcb883f3bdfb4c24b82b21346a0444309e933ff019d5c3f39"
+  url "https://github.com/ConsenSys/teku.git", using: :git, tag: "21.12.2", revision: "a443c80dd686092e535b1e37c26b8ba50234b223"
```
Comment (SMillerDev) [inline Formula/teku.rb:4]: Can you align the arguments like this. It reads nicer.
```suggestion
  url "https://github.com/ConsenSys/teku.git", 
        tag: "21.12.2",
        revision: "a443c80dd686092e535b1e37c26b8ba50234b223"
```

Comment (ajsutton) [reply Formula/teku.rb]: Done.

**Comment by ajsutton** [comment]
Test failure looks like teku didn't start up fast enough and the curl request was made before it was listening.  That's not a new issue introduced with this change but will likely occur reasonably often.  Ideally instead of the `sleep 10` the `curl` command would just be done in a waiter loop so it retries until it passes (with a timeout of something like 30 seconds). I'm not familiar enough with the homebrew test setup to know how to do that properly though (and it probably should be a separate PR anyway).

**Comment by ajsutton** [comment]
I think this one should be good to merge now.

**Review by SMillerDev — APPROVED** [review]
Thanks @ajsutton  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Pull request closed without merging**